File: App/js/sim/tribeComputer.js
"use strict";

function createTribeComputer(util, naming, connectionComputer) {
    function computeTribes(universe, connections, callback) {
        const tribes = [];
        const orderedConnections = connectionComputer.getOrderedConnections(connections);
        universe.forEach(ind => setMembership(ind, null, null));
        orderedConnections.forEach(c => {
            var first = universe[c.x];
            var second = universe[c.y];
            if (!first.tribe) {
                if (!second.tribe) {
                    createTribe([first, second]);
                }
                else {
                    tryJoinTribe(first, second.tribe);
                }
            }
            else {
                if (!second.tribe) {
                    tryJoinTribe(second, first.tribe);
                }
                else if (first.tribe !== second.tribe) {
                    tryMergeTribes(first, second);
                }
            }
        });
        tribes.forEach((t, i) => t.index = i);
        tribes.sort((a, b) => getDominantTribe(a, b) === a ? -1 : +1);
        callback(tribes);

        function areConnected(first, second) {
            return first.index < second.index
                ? areConnected(second, first)
                : connections[first.index][second.index] > 0;
        }

        function canJoinTribe(aspiringMember, tribe) {
            return tribe.members.every(member => areConnected(aspiringMember, member));
        }

        function tryJoinTribe(aspiringMember, tribe) {
            if (!canJoinTribe(aspiringMember, tribe))
                return;
            tribe.members.push(aspiringMember);
            setMembership(aspiringMember, tribe, tribe.members.length);
        }

        function tryMergeTribes(first, second) {
            const dominantTribe = getDominantTribe(first.tribe, second.tribe);
            const secondaryTribe = dominantTribe === first.tribe ? second.tribe : first.tribe;
            var allMembers = dominantTribe.members.concat(secondaryTribe.members);
            if (!canFormOneTribe(allMembers))
                return;
            util.removeFrom(tribes, first.tribe);
            util.removeFrom(tribes, second.tribe);
            createTribe(allMembers);
        }

        function canFormOneTribe(aspiringMembers) {
            var indices = aspiringMembers.map(m => m.index);
            indices.sort((a, b) => a - b);
            const n = indices.length;
            for (var x = 1; x < n; x++)
                for (var y = 0; y < x; y++) {
                    if (!connections[indices[x]][indices[y]])
                        return false;
                }
            return true;
        }

        function getDominantTribe(tribe1, tribe2) {
            if (tribe1.members.length > tribe2.members.length)
                return tribe1;
            if (tribe2.members.length > tribe1.members.length)
                return tribe2;
            return (getFoundingBond(tribe1) > getFoundingBond(tribe2))
                ? tribe1 : tribe2;
        }

        function getFoundingBond(tribe) {
            const ind1 = tribe.members[0];
            const ind2 = tribe.members[1];
            return connections[ind1.index][ind2.index] || connections[ind2.index][ind1.index];
        }

        function createTribe(members) {
            const tribe = {
                name: naming.generateTribeName(tribes, members[0], members[1]),
                members: members
            };
            members.forEach((m, i) => setMembership(m, tribe, i + 1));
            tribes.push(tribe);
        }
    }

    function setMembership(member, tribe, n) {
        member.tribe = tribe;
        member.membershipNumber = n;
    }

    function getOutsiders(universe) {
        return universe.filter(i => !i.tribe);
    }

    return {
        computeTribes: computeTribes,
        getOutsiders: getOutsiders
    };
}

if (typeof module !== 'undefined') {
    module.exports = createTribeComputer;
}

File: Service/Console/Console/TribeComputer.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Console
{
    public class TribeComputer
    {
        public static Tribe[] ComputeTribes(Individual[] individuals, float[][] connections)
        {
            var tribes = new List<Tribe>();
            foreach (var ind in individuals)
                SetMembership(ind, null, 0);
            foreach (var (x, y, strength) in ConnectionComputer.OrderByDecreasingStrength(connections))
            {
                var first = individuals[x];
                var second = individuals[y];
                if (first.Tribe is null)
                {
                    if (second.Tribe is null)
                    {
                        CreateTribe(first, second);
                    }
                    else
                    {
                        TryJoinTribe(first, second.Tribe);
                    }
                }
                else
                {
                    if (second.Tribe is null)
                    {
                        TryJoinTribe(second, first.Tribe);
                    }
                    else if (first.Tribe != second.Tribe)
                    {
                        TryMergeTribes(first, second);
                    }
                }
            }
            for (var i = 0; i < tribes.Count; i++)
                tribes[i].Index = i;
            return tribes.OrderByDescending(GetDominance).ToArray();

            bool AreConnected(Individual first, Individual second)
                => first.Index < second.Index
                    ? AreConnected(second, first)
                    : connections[first.Index][second.Index] > 0;

            bool CanJoinTribe(Individual aspiringMember, Tribe tribe)
                => tribe.Members.All(member => AreConnected(aspiringMember, member));

            void TryJoinTribe(Individual aspiringMember, Tribe tribe)
            {
                if (!CanJoinTribe(aspiringMember, tribe))
                    return;
                tribe.Members.Add(aspiringMember);
                SetMembership(aspiringMember, tribe, tribe.Members.Count);
            }

            void TryMergeTribes(Individual first, Individual second)
            {
                var dominantTribe = GetDominantTribe(first.Tribe, second.Tribe);
                var secondaryTribe = dominantTribe == first.Tribe ? second.Tribe : first.Tribe;
                var allMembers = dominantTribe.Members.Concat(secondaryTribe.Members).ToArray();
                if (!CanFormOneTribe(allMembers))
                    return;
                tribes.Remove(first.Tribe);
                tribes.Remove(second.Tribe);
                CreateTribe(allMembers);
            }

            bool CanFormOneTribe(Individual[] aspiringMembers)
            {
                var indices = aspiringMembers
                    .Select(m => m.Index).ToArray()
                    .OrderBy(i => i)
                    .ToArray();
                var n = indices.Length;
                for (var x = 1; x < n; x++)
                    for (var y = 0; y < x; y++)
                        if (connections[indices[x]][indices[y]] == 0)
                            return false;
                return true;
            }

            (int, float) GetDominance(Tribe tribe) => (tribe.Members.Count, tribe.FoundingBond);

            Tribe GetDominantTribe(Tribe tribe1, Tribe tribe2)
            {
                if (tribe1.Members.Count > tribe2.Members.Count)
                    return tribe1;
                if (tribe2.Members.Count > tribe1.Members.Count)
                    return tribe2;
                return tribe1.FoundingBond > tribe2.FoundingBond ? tribe1 : tribe2;
            }

            float GetFoundingBond(params Individual[] members)
            {
                var ind1 = members[0];
                var ind2 = members[1];
                return Math.Max(connections[ind1.Index][ind2.Index], connections[ind2.Index][ind1.Index]);
            }

            void CreateTribe(params Individual[] members)
            {
                var tribe = new Tribe
                {
                    Name = $"{members[0].Id}-{members[1].Id}",
                    Members = members.ToList(),
                    FoundingBond = GetFoundingBond(members)
                };
                for (var i = 0; i < members.Length; i++)
                    SetMembership(members[i], tribe, i + 1);
                tribes.Add(tribe);
            }
        }

        private static void SetMembership(Individual member, Tribe tribe, int n)
        {
            member.Tribe = tribe;
            member.MembershipNumber = n;
        }
    }
}